actor/root_context.go

```diff
@@ -0,0 +1,116 @@
+package actor
+
+import "time"
+
+type RootContext struct {
+	senderMiddleware SenderFunc
+	headers          messageHeader
+}
+
+var emptyRootContext = &RootContext{
+	senderMiddleware: nil,
+	headers:          emptyMessageHeader,
+}
+
+func EmptyRootContext() *RootContext {
+	return emptyRootContext
+}
+
+func NewRootContext(header map[string]string, middleware ...SenderMiddleware) *RootContext {
+	return &RootContext{
+		senderMiddleware: makeSenderMiddlewareChain(middleware, func(_ SenderContext, target *PID, envelope *MessageEnvelope) {
+			target.sendUserMessage(envelope)
+		}),
+		headers: messageHeader(header),
+	}
+}
+
+func (rc *RootContext) WithHeaders(headers map[string]string) *RootContext {
+	rc.headers = headers
+	return rc
+}
+
+func (rc *RootContext) WithSenderMiddleware(middleware ...SenderMiddleware) *RootContext {
+	rc.senderMiddleware = makeSenderMiddlewareChain(middleware, func(_ SenderContext, target *PID, envelope *MessageEnvelope) {
+		target.sendUserMessage(envelope)
+	})
+	return rc
+}
+
+//
+// Interface: SenderContext
+//
+
+func (rc *RootContext) Message() interface{} {
+	return nil
+}
+
+func (rc *RootContext) MessageHeader() ReadonlyMessageHeader {
+	return rc.headers
+}
+
+func (rc *RootContext) Send(pid *PID, message interface{}) {
+	rc.sendUserMessage(pid, message)
+}
+
+func (rc *RootContext) Request(pid *PID, message interface{}) {
+	rc.sendUserMessage(pid, message)
+}
+
+func (rc *RootContext) RequestWithCustomSender(pid *PID, message interface{}, sender *PID) {
+	env := &MessageEnvelope{
+		Header:  nil,
```

Comment:
Shouldn't this be this?:

```suggestion
		Header:  rc.headers,
```

Reply:
@emilingerslev Hi, thank you for pointing out! I guess the problem here is due to a lack of detailed documentation, and I do agree that it is indeed a bit confusing and not user friendly at this part.

The `nil` here is just to follow the common context pattern, where `context.headers` is  reserved for the incoming message, while for assigning headers in outgoing message, it can only be done in sender middleware, not through context or actor.

This is also the reason why `headers` and `message` field are in `SenderContext` instead of `ReceiverContext`. Since in some sender middleware,  user need to access received message's header like tracing id, etc. 

Hope I explained it clearly! = )

Reply:
Alright, makes sense is just my understanding that's a bit... off 😄 So to clarify; what are the RootContext headers for then?